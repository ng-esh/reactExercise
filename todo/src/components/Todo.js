import React, { useState } from "react";
import "../styles/Todo.scss";

function Todo({ id, task, removeTodo, updateTodo, toggleComplete, isCompleted }) {
  const [isEditing, setIsEditing] = useState(false);
  const [updatedTask, setUpdatedTask] = useState(task);

  const handleUpdate = (e) => {
    e.preventDefault();
    updateTodo(id, updatedTask);
    setIsEditing(false);
  };

  return (
    <li className={`todo-item ${isCompleted ? "completed" : ""}`} data-testid="todo-item">
      {isEditing ? (
        <form onSubmit={handleUpdate} className="edit-form">
          <input
            type="text"
            value={updatedTask}
            onChange={(e) => setUpdatedTask(e.target.value)}
            data-testid="edit-input"
          />
          <button type="submit" data-testid="save-btn">Save</button>
        </form>
      ) : (
        <>
          <span data-testid="todo-text">{task}</span>
          <button className="edit-btn" onClick={() => setIsEditing(true)} data-testid="edit-btn">
            Edit
          </button>
          <button className="remove-btn" onClick={() => removeTodo(id)} data-testid="remove-todo-btn">
            X
          </button>
          <button className="complete-btn" onClick={() => toggleComplete(id)} data-testid="complete-btn">
            {isCompleted ? "Undo" : "Mark as completed"}
          </button>
        </>
      )}
    </li>
  );
}




export default Todo;