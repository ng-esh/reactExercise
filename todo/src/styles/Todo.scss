@import "./animations.scss";

.todo-item {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 8px;
  border: 1px solid #ddd;
  border-radius: 4px;
  margin-bottom: 5px;
  background: white;
  transition: opacity 0.3s ease-in-out;

  &.completed {
    text-decoration: line-through;
    color: gray;
  }

  .edit-form {
    display: flex;
    gap: 5px;

    input {
      flex-grow: 1;
      padding: 5px;
      border: 1px solid #ccc;
      border-radius: 4px;
    }

    button {
      padding: 5px 10px;
      background: #28a745;
      color: white;
      border: none;
      cursor: pointer;
      border-radius: 4px;
      transition: background 0.3s;

      &:hover {
        background: #218838;
      }
    }
  }

  button {
    cursor: pointer;
    padding: 5px;
    border-radius: 4px;
    transition: opacity 0.2s;

    &.remove-btn {
      background: red;
      color: white;
      border: none;

      &:hover {
        background: darkred;
      }
    }

    &.edit-btn {
      background: #ffc107;
      color: black;
      border: none;

      &:hover {
        background: #e0a800;
      }
    }

    &.complete-btn {
      background: #17a2b8;
      color: white;
      border: none;

      &:hover {
        background: #138496;
      }
    }
  }

  &.fade-out {
    opacity: 0;
    transition: opacity 0.5s ease-out;
  }
}
