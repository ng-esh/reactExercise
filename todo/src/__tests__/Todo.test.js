import { render, fireEvent, screen } from "@testing-library/react";
import Todo from "../components/Todo.js";

test("renders without crashing", () => {
  render(<Todo id={1} task="Test Task" removeTodo={() => {}} updateTodo={() => {}} toggleComplete={() => {}} />);
});

test("matches snapshot", () => {
  const { asFragment } = render(<Todo id={1} task="Test Task" removeTodo={() => {}} updateTodo={() => {}} toggleComplete={() => {}} />);
  expect(asFragment()).toMatchSnapshot();
});

test("removes todo when button is clicked", () => {
  const mockRemove = jest.fn();
  render(<Todo id={1} task="Test Task" removeTodo={mockRemove} updateTodo={() => {}} toggleComplete={() => {}} />);
  fireEvent.click(screen.getByTestId("remove-todo-btn"));
  expect(mockRemove).toHaveBeenCalledWith(1);
});

test("edits todo when edit button is clicked", () => {
  const mockUpdate = jest.fn();
  render(<Todo id={1} task="Test Task" removeTodo={() => {}} updateTodo={mockUpdate} toggleComplete={() => {}} />);
  
  fireEvent.click(screen.getByTestId("edit-btn"));
  fireEvent.change(screen.getByTestId("edit-input"), { target: { value: "Updated Task" } });
  fireEvent.click(screen.getByTestId("save-btn"));

  expect(mockUpdate).toHaveBeenCalledWith(1, "Updated Task");
});

test("marks todo as completed", () => {
  const mockToggle = jest.fn();
  render(<Todo id={1} task="Test Task" removeTodo={() => {}} updateTodo={() => {}} toggleComplete={mockToggle} />);
  fireEvent.click(screen.getByTestId("complete-btn"));
  expect(mockToggle).toHaveBeenCalledWith(1);
});
