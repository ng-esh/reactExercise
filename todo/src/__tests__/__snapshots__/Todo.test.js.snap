// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <li
    class="todo-item "
    data-testid="todo-item"
  >
    <span
      data-testid="todo-text"
    >
      Test Task
    </span>
    <button
      class="edit-btn"
      data-testid="edit-btn"
    >
      Edit
    </button>
    <button
      class="remove-btn"
      data-testid="remove-todo-btn"
    >
      X
    </button>
    <button
      class="complete-btn"
      data-testid="complete-btn"
    >
      Mark as completed
    </button>
  </li>
</DocumentFragment>
`;
